src/query/api/v1/handler/json/write.go

```diff
@@ -0,0 +1,115 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3/src/query/util/logging"
+	xtime "github.com/m3db/m3x/time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	// WriteJSONURL is the url for the write json handler
+	WriteJSONURL = handler.RoutePrefixV1 + "/write_json"
+
+	// JSONWriteHTTPMethod is the HTTP method used with this resource.
+	JSONWriteHTTPMethod = http.MethodPost
+)
+
+// WriteJSONHandler represents a handler for the write json endpoint
+type WriteJSONHandler struct {
+	store storage.Storage
+}
+
+// NewWriteJSONHandler returns a new instance of handler.
+func NewWriteJSONHandler(store storage.Storage) http.Handler {
+	return &WriteJSONHandler{
+		store: store,
+	}
+}
+
+// WriteQuery represents the write request from the user
+// NB(braskin): support only writing one datapoint for now
+type WriteQuery struct {
+	Tags      map[string]string `json:"tags" validate:"nonzero"`
+	Timestamp int               `json:"timestamp" validate:"nonzero"`
+	Value     float64           `json:"value" validate:"nonzero"`
+}
+
+func (h *WriteJSONHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	req, rErr := h.parseRequest(r)
+	if rErr != nil {
+		handler.Error(w, rErr.Inner(), rErr.Code())
+		return
+	}
+
+	writeQuery := newStorageWriteQuery(req)
+
+	if err := h.store.Write(r.Context(), writeQuery); err != nil {
+		logging.WithContext(r.Context()).Error("Write error", zap.Any("err", err))
+		handler.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+}
+
+func newStorageWriteQuery(req *WriteQuery) *storage.WriteQuery {
+	return &storage.WriteQuery{
+		Tags: req.Tags,
+		Datapoints: ts.Datapoints{
+			{
+				Timestamp: time.Unix(int64(req.Timestamp), 0),
+				Value:     req.Value,
+			},
+		},
+		Unit:       xtime.Millisecond,
+		Annotation: nil,
+	}
+}
+
+func (h *WriteJSONHandler) parseRequest(r *http.Request) (*WriteQuery, *handler.ParseError) {
```

Comment:
how about supporting multiple write requests in one query ? Can be useful for benchmarks

Reply:
So basically we'd unmarshal into a `[]WriteQuery`?

Reply:
I'm not sure how useful it would be for benchmarks; it's different from the prom endpoint, and if people are running standalone Query in the future, they'd be putting data either directly into m3db or through aggregators, not through this endpoint.